Check for missing WPA CRD before starting WPA controller loop

xornivore opened the pull request:
### What does this PR do?

Wait for WPA CRD infinitely if it is not registered before entering WPA controller loop. 
- Prevents frequent error logs from the Informer trying to list WPA resources.

This uses exponential backoff to increase delay between WPA CRD checks.

### Motivation

If you rollout the DCA without the CRD declared, you will see the following rather frequent log message:

```
E1001 16:14:05.575077       1 reflector.go:125] github.com/DataDog/watermarkpodautoscaler/pkg/client/informers/externalversions/factory.go:106: Failed to list *v1alpha1.WatermarkPodAutoscaler: the server could not find the requested resource (get watermarkpodautoscalers.datadoghq.com)
```

### Additional Notes

Anything else we should know when reviewing?

Commit:
Wait for WPA CRD to become available before starting WPA controller loop

Commit:
Shorten filename

vboulineau commented on pkg/util/kubernetes/apiserver/wpa_controller.go:
```diff
@@ -38,8 +54,62 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
+		Clock:               backoff.SystemClock,
+	}
+	exp.Reset()
+	_ = backoff.Retry(checkWPACRD(wpaClient), exp)
```
Is there a way to get the next "wait" time, so we can display it in the logs? Would be useful if people want to know if stuff is dead or just waiting for retry

xornivore replied:
Ah yes, we could do that, though worth mentioning that we set the max of 	`crdCheckMaxInterval     = 5 * time.Minute` for this retrier.

Commit:
Address review feedback

- Add the next retry delay time in log trace.

Commit:
Remove delay randomization, clean up log trace

clamoriniere commented on pkg/util/kubernetes/apiserver/wpa_controller.go:
```diff
@@ -38,8 +55,69 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
+
+func checkWPACRD(wpaClient wpa_client.Interface) backoff.Operation {
+	check := func() error {
+		_, err := wpaClient.DatadoghqV1alpha1().WatermarkPodAutoscalers(v1.NamespaceAll).List(v1.ListOptions{})
```
maybe you can use the `CustomResourceDefinition` client here.
something like:
```
client.ApiextensionsV1beta1().CustomResourceDefinitions().Get(...)
```

xornivore replied:
Unfortunately the `clientSet` in WPA controller does not have `ApiextensionsV1beta()` API set.

clamoriniere commented on pkg/util/kubernetes/apiserver/wpa_controller.go:
```diff
@@ -38,8 +55,69 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
+}
+
+func waitForWPACRD(wpaClient wpa_client.Interface) {
+	exp := &backoff.ExponentialBackOff{
```
did you look at kubernetes internal for backoff implementation: `wait.UntilWithContext `

```https://github.com/kubernetes/apimachinery/blob/9a93b3ad87694c17927348f74a0d8742ffa6221c/pkg/util/wait/wait.go#L96
```

xornivore replied:
Thank you @clamoriniere, I looked at the `ExponentialBackoff`, it's pretty nice but unfortunately it doesn't have a couple of convenient features:
a) it doesn't have a _clean way_ to enable infinite retries.
b) it doesn't have a way to retrieve the delay for the next attempt.

clamoriniere reviewed (commented):
looks great ! 
Maybe we can remove the new dependency by using internal kubernetes helper

Commit:
Fix the problem with Informer initialization for WPA

- Move start of the informer to the WPA check goroutine to fix the problem.
- Cleanup and code review feedback
- TODO for later to refactor the shared informer factory initialization.

Commit:
Remove a sub-package from Gopkg.lock

clamoriniere reviewed (approved):
👍
